### PR TITLE
Fixing bugs of blocker and major for module JTS for GF

### PR DESCRIPTION
**Pull request by svendiedrichsen**
https://sonarcloud.io/project/issues?id=org.glassfish.main%3Aglassfish-main-aggregator&moduleUuids=AWJyK9gx1B6WA4TgV3Cw&resolved=false&types=BUG

### CODE REVIEW

**Comment by svendiedrichsen** [comment]
Does it make sense to refactor code in the package `com.sun.jts.codegen` and below or will it be regenerated at some point?

File: appserver/transaction/jts/src/main/java/com/sun/jts/CosTransactions/RecoveryManager.java [hunk 1]
```diff
@@ -1645,19 +1643,19 @@ public static void recoverIncompleteTx(XAResource[] xaresArray) {
                                                 commitRetriesLeft--;
                                             }
                                             try {
-                                                Thread.sleep(Configuration.COMMIT_RETRY_WAIT);
+                                                inCompleteTxMap.wait(Configuration.COMMIT_RETRY_WAIT);
```
Comment (smillidge) [inline appserver/transaction/jts/src/main/java/com/sun/jts/CosTransactions/RecoveryManager.java]: this change may change behavior in the core TX manager.

File: appserver/transaction/jts/src/main/java/com/sun/jts/CosTransactions/SubCoordinator.java [hunk 2]
```diff
@@ -1738,7 +1736,7 @@ public void getResources(ResourceSequenceHolder resources,
      *
      * @see
      */
-    CompletionHandler getTerminator() {
+    synchronized CompletionHandler getTerminator() {
```
Comment (smillidge) [inline appserver/transaction/jts/src/main/java/com/sun/jts/CosTransactions/SubCoordinator.java]: this could change behaviour what is the rationale behind this change?

File: appserver/transaction/jts/src/main/java/com/sun/jts/codegen/jtsxa/OTSResourceHelper.java [hunk 3]
```diff
@@ -53,13 +60,19 @@
...
   private static String  _id = "IDL:jtsxa/OTSResource:1.0";
+  private static final Logger LOGGER = Logger.getLogger(OTSResourceHelper.class.getName());
 
   public static void insert (org.omg.CORBA.Any a, com.sun.jts.codegen.jtsxa.OTSResource that)
   {
-    org.omg.CORBA.portable.OutputStream out = a.create_output_stream ();
-    a.type (type ());
-    write (out, that);
-    a.read_value (out.create_input_stream (), type ());
+    try (org.omg.CORBA.portable.OutputStream out = a.create_output_stream ()) {
```
Comment (smillidge) [inline appserver/transaction/jts/src/main/java/com/sun/jts/codegen/jtsxa/OTSResourceHelper.java]: Are you sure these are output and input streams that need closing here?

File: appserver/transaction/jts/src/main/java/com/sun/jts/codegen/otsidl/CoordinatorResourceHelper.java [hunk 4]
```diff
@@ -53,13 +60,19 @@
...
   private static String  _id = "IDL:otsidl/CoordinatorResource:1.0";
+  private static final Logger LOGGER = Logger.getLogger(CoordinatorResourceHelper.class.getName());
 
   public static void insert (org.omg.CORBA.Any a, com.sun.jts.codegen.otsidl.CoordinatorResource that)
   {
-    org.omg.CORBA.portable.OutputStream out = a.create_output_stream ();
-    a.type (type ());
-    write (out, that);
-    a.read_value (out.create_input_stream (), type ());
+    try (org.omg.CORBA.portable.OutputStream out = a.create_output_stream ()) {
```
Comment (smillidge) [inline appserver/transaction/jts/src/main/java/com/sun/jts/codegen/otsidl/CoordinatorResourceHelper.java]: Same comment as above

File: appserver/transaction/jts/src/main/java/com/sun/jts/codegen/otsidl/JControlHelper.java [hunk 5]
```diff
@@ -53,13 +60,19 @@
...
   private static String  _id = "IDL:otsidl/JControl:1.0";
+  private static final Logger LOGGER = Logger.getLogger(JControlHelper.class.getName());
 
   public static void insert (org.omg.CORBA.Any a, com.sun.jts.codegen.otsidl.JControl that)
   {
-    org.omg.CORBA.portable.OutputStream out = a.create_output_stream ();
-    a.type (type ());
-    write (out, that);
-    a.read_value (out.create_input_stream (), type ());
+    try (org.omg.CORBA.portable.OutputStream out = a.create_output_stream ()) {
```
Comment (smillidge) [inline appserver/transaction/jts/src/main/java/com/sun/jts/codegen/otsidl/JControlHelper.java]: Same comment as above

File: appserver/transaction/jts/src/main/java/com/sun/jts/codegen/otsidl/JCoordinatorHelper.java [hunk 6]
```diff
@@ -53,13 +60,19 @@
...
   private static String  _id = "IDL:otsidl/JCoordinator:1.0";
+  private static final Logger LOGGER = Logger.getLogger(JCoordinatorHelper.class.getName());
 
   public static void insert (org.omg.CORBA.Any a, com.sun.jts.codegen.otsidl.JCoordinator that)
   {
-    org.omg.CORBA.portable.OutputStream out = a.create_output_stream ();
-    a.type (type ());
-    write (out, that);
-    a.read_value (out.create_input_stream (), type ());
+    try (org.omg.CORBA.portable.OutputStream out = a.create_output_stream ()) {
```
Comment (smillidge) [inline appserver/transaction/jts/src/main/java/com/sun/jts/codegen/otsidl/JCoordinatorHelper.java]: Same comment as above

**Comment by smillidge** [comment]
Thanks for doing this work. I have some comments where I wonder whether behaviour is being changed. I admit I haven't analysed the code to see of these are valid concerns.

File: appserver/transaction/jts/src/main/java/com/sun/jts/codegen/jtsxa/OTSResourceHelper.java (hunk 3 above)
Comment (svendiedrichsen) [reply appserver/transaction/jts/src/main/java/com/sun/jts/codegen/jtsxa/OTSResourceHelper.java]: Closing those streams will release ByteBuffers acquired more quickly. Will probably reduce memory footprint.

File: appserver/transaction/jts/src/main/java/com/sun/jts/codegen/otsidl/CoordinatorResourceHelper.java (hunk 4 above)
Comment (svendiedrichsen) [reply appserver/transaction/jts/src/main/java/com/sun/jts/codegen/otsidl/CoordinatorResourceHelper.java]: See above.

File: appserver/transaction/jts/src/main/java/com/sun/jts/codegen/otsidl/JControlHelper.java (hunk 5 above)
Comment (svendiedrichsen) [reply appserver/transaction/jts/src/main/java/com/sun/jts/codegen/otsidl/JControlHelper.java]: See above.

File: appserver/transaction/jts/src/main/java/com/sun/jts/codegen/otsidl/JCoordinatorHelper.java (hunk 6 above)
Comment (svendiedrichsen) [reply appserver/transaction/jts/src/main/java/com/sun/jts/codegen/otsidl/JCoordinatorHelper.java]: See above.

File: appserver/transaction/jts/src/main/java/com/sun/jts/CosTransactions/SubCoordinator.java (hunk 2 above)
Comment (svendiedrichsen) [reply appserver/transaction/jts/src/main/java/com/sun/jts/CosTransactions/SubCoordinator.java]: To mirror the synchronization from the `setTerminator` Method. Please see https://www.securecoding.cert.org/confluence/x/I4BoAg

**Comment by arjantijms** [comment]
Jenkins test please

**Comment by arjantijms** [comment]
Jenkins test please

File: appserver/transaction/jts/src/main/java/com/sun/jts/codegen/jtsxa/OTSResourceHelper.java (hunk 3 above)
Comment (MattGill98) [reply appserver/transaction/jts/src/main/java/com/sun/jts/codegen/jtsxa/OTSResourceHelper.java]: I think closing this looks like a positive change, so I'm happy with it. Obviously it still needs a second reviewer though, so ping!
